docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
 
 is diagnosed as a label in which the user forgot the **label** keyword.
 
+## Update with Failure Statement
+
+An update statement using `:-`   instead of `:=` exits execution of a method immediately if a failure is reported. This is a form of exceptional return from the calling method.
+
+To use this form of update,
+   * the caller must have a first output value whose type is failure-compatible
+   * if the RHS of the update-with-failure statement is a method call, the first output value of the callee must be failure-compatible
+   * if instead the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
+
+A failure-compatible type is a type that has the following:
+   * a predicate `predicate IsFailure()`
+   * a method `method PropagateFailure() returns _X_ `: The type _X_ must be a failure-compatible type and must be the type of the first output of the caller.
+   * a method `method Extract() returns _T_` where _T_ is some type for the value carried by the type.
+
+TODO: Why should PropagateFailure and Extract be functions?
+
+### RHS with method call
+
+A common example of a failure-compatible type is the following:
+```
+datatype Outcome<T> =
+| Success(value: T)
+| Failure(error: string)
+{
+  predicate method IsFailure() {
+    this.Failure?
+  }
+  function method PropagateFailure(): Outcome<U>
+    requires IsFailure()
+  {
+    Failure(this.error); // this is Outcome<U>.Failure(...)
+  }
+  function method Extract(): T
+    requires !IsFailure()
+  {
+    this.value
+  }
+}
+```
+
+The use of this statement is best explained by an example:
+```
+method callee(i: int) returns (r: Outcome<nat>, v: int)
+{
+  if (i < 0) return Failure("negative")
+  return Success(i), i+i;
+}
+
+method caller(i: int) returns (rr: Outcome<int>, k: int)
+{
+  var j: int;
+  j, k :- callee(i);
+  k := k + k;
+}
+```
+Suppose `caller` is called with an argument of `10`. Then `callee` is called with argument `10` and returns `r` and `v` of `Outcome<nat>.Success(10)` and `20`. Here `r.IsFailure()` is `false`, so control proceeds normally. The `j` is assigned the result of `r.Extract()`, which will be `10`, and `k` is assigned `20`. Control flow proceeds to the next line, where `k` now gets the value `40`.
+
+Suppose instead  `caller` is called with an argument of `-1`. Then `callee` is called with the value `-1` and returns `r` and `v` of `Outcome<nat>.Failure("negative")` and `-2`. Now `r.IsFailure()` is `true`, so control proceeds directly to return from `caller`. The output values of `caller`  (`rr` and `k`) get the values of `r.PropagateFailure()`, which is `Outcome<int>.Failure("negative")`, and `-2`. The rest of the body of `caller` is skipped. In addition, the first return value of `caller` is now a failure value, so the exceptional return will propagate up the call stack as long as there are callers with this first special output type and calls that use `:-`.
```

Comment:
L304:

    The first out-parameter of `Caller` (`rr`) gets its value from `r.PropagateFailure()`.
    The other out-parameter (`k`) is not changed by the `:-` statement, so it will return from
    `Caller` with the value it has, which in the example is an arbitrary value.

It may be clearer to assign `k` (say, setting it to `5`) in `Caller` before the `:-` statement. Then, the text I proposed here will say

    The other out-parameter (`k`) is not changed by the `:-` statement, namely `5`.

Reply:
Also, L307 through L309 are true for this example, but doesn't apply if the `Caller` out-parameter had a different type. Therefore, I'd prefer to insert "`In this example,`" or something in L307.

Reply:
I believe that in r, k :- ..., k is assigned whether or not the failure control path is taken.